Unified write_var calls and automated writing variables

singhbalwinder opened the pull request:
Unified write_var calls so that we have the same call signature for all variables irrespective of their dimensions (1D, 2D, etc.). The process of creating test files for each subroutine is now automated where the calls for writing intent/in/outs variables are pre-populated in the files capturing input and outputs.

A major YAML I/O F90 file reorg where new unified call signatures are introduced while maintaining backward compatibility.

singhbalwinder committed:
Adds a script for writing automated code writing for in-out calls and unified yaml i/o interface

singhbalwinder committed:
Adds new python scripts for adding code for calling ins-outs

tangshuaiqi commented on mam4_refactor_scripts/beg.inp:
```diff
@@ -1,69 +1,92 @@
+
+     !-----------------------------------------------------------------------------------------
+     !In the case of y_i or y_k are not passed as arguments, use the following if condition:
      !if(yaml%col_print >0 .and. y_nstep==yaml%nstep_print) then
```
maybe reduce the indent to the previous line so we know that they are alternate options of the above if condition, other than an inner loop.

tangshuaiqi commented on mam4_refactor_scripts/beg.inp:
```diff
@@ -1,69 +1,92 @@
+     if (n_calls==1) then ! output at the first call only, modify this if condition (see below) if writing out other calls
 
-        !open I/O yaml files (it can have an extra optional argument to pass a unique string to differentiate file names)
+        !if ((n_calls==1 .and. flag==0) .or. (n_calls==3 .and. flag==1) .or. (n_calls==5 .and. flag==2)) then
```
reduce the indent of this line too

tangshuaiqi reviewed (approved):
looks great. I only have a minor comments on the indent of alternative if conditions.

brigaudet commented on mam4_refactor_scripts/test_gen_for_sub.sh:
```diff
@@ -73,15 +87,39 @@ main() {
+
 create_file () {
 
     #create_file $module_name $sub_name $yaml_dir_path $relative_path "end"
```
In above comment could change ' "end" ' to ' "beginning"/"end" '.

brigaudet replied:
Actually, I guess ' $ftmp_ins/$ftmp_outs ' also would need to be inserted after ' $relative_path '.

brigaudet reviewed (approved):
Looks good, with only one comment.

singhbalwinder committed:
Fixes indents for if condition and fix comments about function arguments

singhbalwinder committed:
Fixes another indent for an if condition

singhbalwinder commented:
Thanks, guys for a quick review! I have addressed your comments.

FYI @tangshuaiqi and @brigaudet: The new scripts automatically add the `write_var` call in the *_beg and *_end files. We don't need to copy-paste these calls anymore.

singhbalwinder commented:
All tests are PASS:
$ /compyfs/sing201/e3sm_scratch/yam_io_scr1/cs.status.comp_yam_io_scr1_mam4_org_v2_baselines -s
comp_yam_io_scr1_mam4_org_v2_baselines: 9 tests
  PASS ERP_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS ERS.ne4pg2_oQU480.F2010.compy_intel
  PASS PEM_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS PET.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D_Ln5_P1x1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D_Ln5_P32x1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_Ly1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS.ne30pg2_EC30to60E2r2.F2010.compy_intel
